cpp/src/replace/replace.cu

```diff
@@ -53,16 +53,27 @@ namespace{ //anonymous
                       gdf_size_type                      nrows,
                       thrust::device_ptr<const T> old_values_begin,
                       thrust::device_ptr<const T> old_values_end,
-                      const T*                    d_new_values)
+                      const T*                    d_new_values,
+                      gdf_valid_type*             col_valid,
+                      gdf_valid_type*             new_valid)
   {
     gdf_size_type i = blockIdx.x * blockDim.x + threadIdx.x;
     while(i < nrows)
     {
-      auto found_ptr = thrust::find(thrust::seq, old_values_begin, old_values_end, d_col_data[i]);
-
-      if (found_ptr != old_values_end) {
-          auto d = thrust::distance(old_values_begin, found_ptr);
-          d_col_data[i] = d_new_values[d];
+      if ( gdf_is_valid(col_valid, i)){
+          auto found_ptr = thrust::find(thrust::seq, old_values_begin, old_values_end, d_col_data[i]);
+
+          if (found_ptr != old_values_end) {
+              auto d = thrust::distance(old_values_begin, found_ptr);
+              if (gdf_is_valid(new_valid, d)){
+                d_col_data[i] = d_new_values[d];
+              }
+              else{
+                //unset the i-th bit in col_valid
+                gdf_unset_bit(col_valid, i);
+              }
+
```

Comment:
Just a comment for the record: there's probably an opportunity to optimize this kernel by staging the data in shared memory and writing out coalesced without atomics.  Depends on the number of values that match the to_replace though. If it's small, this will be fine.

Reply:
This sounds good. I can definitely look for this optimization.